refactor(forms): allow passing number|string|null paths to min & max

mmalerba opened the pull request:
Relaxes the constraints on which paths can be used with the `min` & `max` validation rules, since people may want to validate a potentially-null number, or a numeric value represented as a string

Fixes #63789

mmalerba committed:
refactor(forms): allow passing number|string|null paths to min & max

Relaxes the constraints on which paths can be used with the `min` &
`max` validation rules, since people may want to validate a
potentially-null number, or a numeric value represented as a string

kirjs commented on packages/forms/signals/src/api/validators/max.ts:
```diff
@@ -46,7 +46,9 @@ export function max<TPathKind extends PathKind = PathKind.Root>(
     }
-    if (ctx.value() > max) {
+    const value = ctx.value();
+    const numValue = !value && value !== 0 ? NaN : Number(value); // Treat `''` and `null` as `NaN`
```
https://github.com/angular/angular/blob/de234def0e3b0e1ffb3f1ac79d5e9ddeb140c3f7/packages/forms/src/validators.ts#L470
we should probably use parseFloat to be consistent with reactive forms

mmalerba replied:
I don't know, `parseFloat` has weird behavior, it treats `12shdgdhgds` as `12`

kirjs replied:
I haven't see people complaining about that. 
It would also handle the cases like `5.` better, or scientific notation which is (unfortunately?)  accepted by input type number, and is basically more consistent with HTML rules

mmalerba replied:
```
parseFloat('5.') // => 5
Number('5.') // => 5
parseFloat('1e1') // => 10
Number('1e1') // => 10
```

I have not seen any case where we would prefer the `parseFloat` behavior

mmalerba committed:
fixup! refactor(forms): allow passing number|string|null paths to min & max

mmalerba committed:
refactor(forms): improve typing on min & max

If we're calling `min` on a path that's guaranteed to be `number` we
don't want to make the users validator function handle the `null` or
`string` cases.

This uncovered an issue in the `SchemaTreePath` type which needed to be
fixed by preventing the model type from being distributed over.

thePunderWoman reviewed (approved):
reviewed-for: public-api
